### ISSUE TITLE
Filter return type erased when filter function's input type declared to be object

### ISSUE THREAD

**Issue by layday** [microsoft/pyright#1091]
**Describe the bug**
When filtering certain iterables, the return type of the iterator's members is inferred to be `object`, presumably because that is what the filter function is annotated to take as input.

**To Reproduce**
Type check the snippet below.

**Expected behavior**
That `foo` would be of type `Iterator[int]`.

**Screenshots or Code**
```python
# pyright: strict

def filter_fn(value: object):
    ...

foo = filter(filter_fn, {1: ...})
reveal_type(foo)  # Type of "foo" is "Iterator[object]"
```

**VS Code extension or command-line**
Extension, 1.1.78

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is now addressed in pyright 1.1.87, which I just published. It will also be included in the next release of pylance.
